Automattic/wp-calypso
# Gutenlypso: Disable the Publish button if the email is unverified

#### Changes proposed in this Pull Request

Fixes #29539.

This PR modifies the `PostPublishButtonOrToggle` component from the `edit-post` in order to disable the Publish button if the email of the user is not verified.

Since it is now very different from the original component in the `@wordpress/edit-post` package, I decided to move it `/gutenberg/editor/components` so it's clear we're using a customized version.

Apart from displaying the Publish button, this component now do 3 more things if the email is unverified:
* Dispatches a `lockPostSaving` action, so the Publish button is disabled. Note that if the Pre-publish checks are enabled, only the Publish button from the Pre-publish sidebar will be disabled.
* Displays a warning notice with the same messages we display on the classic editor ([handled by `EditorGroundControl`](https://github.com/Automattic/wp-calypso/blob/2964f6c86eafb1a113531d68e60215bd32644268/client/post-editor/editor-ground-control/index.jsx#L77-L92)).
* Renders the `VerifyEmailDialog` component when the "Learn more" link is clicked.

<img width="1274" alt="screen shot 2018-12-19 at 11 53 35" src="https://user-images.githubusercontent.com/1233880/50216290-a37e2e00-0385-11e9-8768-16870ea3080f.png">

<img width="1276" alt="screen_shot_2018-12-19_at_11_53_46" src="https://user-images.githubusercontent.com/1233880/50216277-96613f00-0385-11e9-966f-c999b4a36988.png">

#### Testing instructions

* Go to http://calypso.localhost:3000/start and create a new user without verifying the email. Don't use an @automattic.com email since they need to be always verified before logging in. You can use a personal Gmail account, even if it's already registered, by appending a `+` suffix (i.e. if your email is example@gmail.com, you can use example+20181219test@gmail.com). WordPress.com will considere it a different email but example@gmail.com will still receive any email sent to example+20181219test@gmail.com).
* Start creating a new post (it should use Gutenlypso by default).
* Make sure that a warning notice appears with the message "To publish, check your email and confirm your address. Learn more"
<img width="453" alt="screen shot 2018-12-19 at 11 12 04" src="https://user-images.githubusercontent.com/1233880/50216797-0de39e00-0387-11e9-82b9-839d3c0e4a03.png">

* Click on Learn more and check that the below dialog appears.
<img width="560" alt="screen_shot_2018-12-19_at_11_53_46" src="https://user-images.githubusercontent.com/1233880/50216878-500cdf80-0387-11e9-87af-4d8251b961dc.png">

* Confirm that the "Publish..." button is not disabled and that after clicking on it, the "Publish" button on the pre-publish sidebar is disabled and you cannot click on it.
<img width="1276" alt="screen shot 2018-12-19 at 12 14 23" src="https://user-images.githubusercontent.com/1233880/50217097-0bce0f00-0388-11e9-85c1-9d387e94c1cc.png">
<img width="1275" alt="screen shot 2018-12-19 at 12 16 33" src="https://user-images.githubusercontent.com/1233880/50217099-0bce0f00-0388-11e9-8240-70306ba59c30.png">

* Change the publish date to some date in the future and make sure that the verification email notice changes to "To schedule, check your email and confirm your address."
![image](https://user-images.githubusercontent.com/1233880/50217395-d4139700-0388-11e9-887c-e9570fc12433.png)

* On the More menu, click on Options and uncheck the "Enable Pre-publish Checks" option.
<img width="419" alt="screen shot 2018-12-19 at 12 23 55" src="https://user-images.githubusercontent.com/1233880/50217451-ff968180-0388-11e9-9acf-39f9e5ef8c4a.png">

* Note that the "Schedule..." button in the header changes to "Schedule" and it is now disabled and you cannot click on it.
<img width="1277" alt="screen shot 2018-12-19 at 12 25 12" src="https://user-images.githubusercontent.com/1233880/50217524-2eacf300-0389-11e9-9e23-c4eb11c03bd7.png">

* Verify now the email, reload the editor and check that no notice is displayed and that the Publish/Schedule buttons are enabled.
